modules/service/src/main/scala-3/higherkindness/mu/rpc/avro/AvroUnions.scala

```diff
@@ -0,0 +1,567 @@
+        case v: T2 => Encoder[T2].encode(AvroSchema[T2])(v)
+        case v: T3 => Encoder[T3].encode(AvroSchema[T3])(v)
+        case v: T4 => Encoder[T4].encode(AvroSchema[T4])(v)
+        case v: T5 => Encoder[T5].encode(AvroSchema[T5])(v)
+        case v: T6 => Encoder[T6].encode(AvroSchema[T6])(v)
+        case v: T7 => Encoder[T7].encode(AvroSchema[T7])(v)
+        case v: T8 => Encoder[T8].encode(AvroSchema[T8])(v)
+      }
+    }
+
+  given [
+    T1: SchemaFor: Decoder: TypeGuardedDecoding,
+    T2: SchemaFor: Decoder: TypeGuardedDecoding,
+    T3: SchemaFor: Decoder: TypeGuardedDecoding,
+    T4: SchemaFor: Decoder: TypeGuardedDecoding,
+    T5: SchemaFor: Decoder: TypeGuardedDecoding,
+    T6: SchemaFor: Decoder: TypeGuardedDecoding,
+    T7: SchemaFor: Decoder: TypeGuardedDecoding,
+    T8: SchemaFor: Decoder: TypeGuardedDecoding
+  ]: Decoder[AvroUnion8[T1, T2, T3, T4, T5, T6, T7, T8]] with
+    def decode(schema: Schema): Any => AvroUnion8[T1, T2, T3, T4, T5, T6, T7, T8] = {
+      val schema1 = AvroSchema[T1]
+      val schema2 = AvroSchema[T2]
+      val schema3 = AvroSchema[T3]
+      val schema4 = AvroSchema[T4]
+      val schema5 = AvroSchema[T5]
+      val schema6 = AvroSchema[T6]
+      val schema7 = AvroSchema[T7]
+      val schema8 = AvroSchema[T8]
+
+      { value =>
+        if (canDecode[T1](schema1)(value)) {
+          val decoded: T1 = Decoder[T1].decode(schema1).apply(value)
+          AvroUnion8[T1, T2, T3, T4, T5, T6, T7, T8](decoded)
+        } else if (canDecode[T2](schema2)(value)) {
+          val decoded: T2 = Decoder[T2].decode(schema2).apply(value)
+          AvroUnion8[T1, T2, T3, T4, T5, T6, T7, T8](decoded)
+        } else if (canDecode[T3](schema3)(value)) {
+          val decoded: T3 = Decoder[T3].decode(schema3).apply(value)
+          AvroUnion8[T1, T2, T3, T4, T5, T6, T7, T8](decoded)
+        } else if (canDecode[T4](schema4)(value)) {
+          val decoded: T4 = Decoder[T4].decode(schema4).apply(value)
+          AvroUnion8[T1, T2, T3, T4, T5, T6, T7, T8](decoded)
+        } else if (canDecode[T5](schema5)(value)) {
+          val decoded: T5 = Decoder[T5].decode(schema5).apply(value)
+          AvroUnion8[T1, T2, T3, T4, T5, T6, T7, T8](decoded)
+        } else if (canDecode[T6](schema6)(value)) {
+          val decoded: T6 = Decoder[T6].decode(schema6).apply(value)
+          AvroUnion8[T1, T2, T3, T4, T5, T6, T7, T8](decoded)
+        } else if (canDecode[T7](schema7)(value)) {
+          val decoded: T7 = Decoder[T7].decode(schema7).apply(value)
+          AvroUnion8[T1, T2, T3, T4, T5, T6, T7, T8](decoded)
+        } else if (canDecode[T8](schema8)(value)) {
+          val decoded: T8 = Decoder[T8].decode(schema8).apply(value)
+          AvroUnion8[T1, T2, T3, T4, T5, T6, T7, T8](decoded)
+        } else {
+          throw decodingException(value, schema1, schema2, schema3, schema4, schema5, schema6, schema7, schema8)
+        }
+      }
+    }
+
```

Comment:
In case you're wondering, this file was not auto-generated, unless you count vim macros 😄